502 his time series new design

zhiyuli opened the pull request:
This PR has addressed issues #708 #635 #502 for HIS Referenced Time Series

hydrobuild commented:
Can one of the admins verify this patch?

pkdash commented:
@selimnairb Please review this.

selimnairb commented on hydroshare/settings.py:
```diff
@@ -529,5 +529,10 @@
+            'handlers':['djangolog'],
+            'propagate': False,
+            'level':'DEBUG',
+        },
```
@zhiyuli There needs to be a good reason to define a new logger for a given resource type.  In my own development, I haven't seen a reason not to use the 'django' logger.

selimnairb commented on ref_ts/models.py:
```diff
@@ -49,304 +31,50 @@ def can_have_multiple_files(cls):
+    name = models.CharField(max_length=500, default="", blank=True)
+    code = models.CharField(max_length=500, default="", blank=True)
+    net_work = models.CharField(max_length=500, default="", blank=True)
+    latitude = models.DecimalField(max_digits=9, decimal_places=6, null=True, blank=True)
```
@zhiyuli Are lat and long needed, or can the geographic coverage be used instead?

horsburgh replied:
The regular time series resource uses the Dublin Core coverage element to express the coordinates of the monitoring site.

selimnairb commented on ref_ts/owslib_revised/__init__.py:
```diff
@@ -0,0 +1 @@
+__author__ = 'hydro'
```
@zhiyuli not needed and not informative as written.

selimnairb commented on ref_ts/owslib_revised/waterml/__init__.py:
```diff
@@ -0,0 +1 @@
+__author__ = 'hydro'
```
@zhiyuli not needed and not informative as written.

selimnairb commented on ref_ts/owslib_revised/waterml/wml.py:
```diff
@@ -0,0 +1,937 @@
+from __future__ import (absolute_import, division, print_function)
```
@zhiyuli So now we're maintaining or own fork of OWSLib?  Who is going to track security updates and apply them over time?  OWSLib is open source, and the developer is good about taking contributions (I have done so).  We should use a specific version of OWSLib and contribute necessary changes if the current version is broken for wml.

selimnairb commented on ref_ts/owslib_revised/waterml/wml10.py:
```diff
@@ -0,0 +1,34 @@
+from __future__ import (absolute_import, division, print_function)
```
@zhiyuli Same as above.  We should not be forking OWSLib.

selimnairb commented on ref_ts/owslib_revised/waterml/wml11.py:
```diff
@@ -0,0 +1,35 @@
+from __future__ import (absolute_import, division, print_function)
```
@zhiyuli Same as above.

selimnairb commented on ref_ts/static/ref_ts/d3/LICENSE:
```diff
@@ -1,26 +0,0 @@
-Copyright (c) 2010-2014, Michael Bostock
```
We should include D3 in hs_core's static files as multiple resources are likely to use it.  Otherwise, we risk having multiple versions from different resources, which may lead to conflicts that are hard to debug.

selimnairb commented on ref_ts/static/ref_ts/datepicker/css/datepicker.css:
```diff
@@ -1,182 +0,0 @@
-/*!
```
@zhiyuli Likewise for datepicker and other JS libraries.  We need to think carefully about how to integrate these.  I would much rather not allow each resource type to import arbitrary JS libraries and accompanying CSS will-nilly.

selimnairb commented on ref_ts/static/ref_ts/rickshaw/.jshintrc:
```diff
@@ -1,11 +0,0 @@
-{
```
@zhiyuli Same with rickshaw.  @mauriel133 @horsburgh How would you prefer we handle JS dependencies?  My gut tells me we should minimize the number of JS libraries that we allow resource instances to import, and try to consolidate JS libraries, as much as possible, within hs_core.

selimnairb commented on ref_ts/static/ref_ts/rickshaw/Makefile:
```diff
@@ -1,95 +0,0 @@
-NODE_PREFIX=$(shell npm prefix)
```
@zhiyuli If we are including JS libraries, surely we don't need to include their makefiles in the hydroshare repo, just the built products.

selimnairb commented on ref_ts/static/ref_ts/rickshaw/README.md:
```diff
@@ -1,207 +0,0 @@
-# Rickshaw
```
@zhiyuli Also, no need to litter up the hydroshare repo (and static files) with readme documents for JS libraries.

selimnairb commented on ref_ts/static/ref_ts/rickshaw/examples/ajax.html:
```diff
@@ -1,44 +0,0 @@
-<!doctype>
```
@zhiyuli ... nor examples ...

selimnairb commented:
@zhiyuli I know you have put a lot of work into this PR, but given that the clean up of un-used JavaScript libraries is obscuring meaningful changes in the pull request, I would prefer it if you make the PR in two stages: First commits related to clean up alone, followed by a second PR containing commits for the substantive changes.  As it stands now, I cannot easily evaluate all of the substantive changes here.

zhiyuli commented:
@selimnairb OK. I will keep this PR open, and will submit another PR-clean that will remove all unused js libraries. Once PR-clean gets merged first, current PR should show the real substantive changes. Thanks

@mjstealey will this work?

selimnairb commented:
@zhiyuli Thanks.  Sorry, I know it is extra work, but I want to make sure we can adequately review the code.

zhiyuli commented:
@selimnairb  you good. I should have though of this before submitting that PR. Thanks for your time reviewing this.

pkdash commented:
@zhiyuli Is this ready for a final review?

zhiyuli commented:
@pkdash this pr removed many unused js from ref_ts. Git tracks those deletion and cant display other chances properly. so I submitted another PR #894 to remove those js first. if you have a chance please review #894 and merge it first. Then this PR will be ready for reviewing. thanks

pkdash commented:
@zhiyuli Since @selimnairb has already done some review of it, it would be better if he does a final review.

zhiyuli commented:
@selimnairb 
The branch that removes all unused js from ref_ts has been merged into dev. Then I merged dev into this PR. Now it looks better. The meaningful changes are shown more clearly.

Will you please continue reviewing it? thanks

About the OWSLib, I added OWSLib==0.10.3 as a dependency to parse wml1 (hydroshare will install it from pypi). But I found it can't parse some wml correctly (maybe the owslib does not support all wml implementations, or maybe the wml returned by some cuahsi hydroservers are not well formatted, anyway the owslib does match some wml in this case). I found that only three python module files in owslib are implementing this parsing function (but the three files are referencing some other OWSLib modules), so I copied/forked these three source files, slightly modified them to fix above issue and included them in my ref_ts source code.

selimnairb commented:
@zhiyuli Placing a modified subset of OWSLib (or any open source library) in HydroShare is problematic for a few reasons:

First, it may be illegal.  You need to very carefully check the terms of the OWSLib [license](https://github.com/geopython/OWSLib/blob/master/LICENSE.txt) and make sure that what you are doing is allowed (it may not be).  At the very least you are not complying with the terms of the OWSLib license as you did not include the license text along with the code that you are redistributing. 

Second, you have benefitted from the OWSLib project, but without contributing code back to the project.

Third, we would be introducing a now-orphaned version of a subset of the OWSLib code into HydroShare.  If security vulnerabilities (remember, you are parsing XML potentially from third party sources that we do not control) or other bugs are fixed in OWSLib, how do we make sure that such changes are ported to our version of the code?

For these reasons, I cannot support using OWSLib in this way.  An acceptable alternative would be to fork OWSLib (as part of the HydroShare organization ideally), fix the bugs that prevent it from doing what you need, and then submit a pull request to OWSLib.  Until this pull request is merged into OWSLib and released, we can run our fork of the entire OWSLib.  @mjstealey or myself can help you with these logistics.  It shouldn't take us too long to do the right thing here.

zhiyuli commented:
@selimnairb Thank you for your comments. I am thinking the changes I made to owslib is to handle some special wml returned by some cuahsi hydroservers. I am guessing these wml do not fully conform to wml standards. so the changes I made to owslib is not of general interest and may not be included in official owslib release. what should we do in this case? thanks

selimnairb commented:
@zhiyuli Can the hydroservers be made to be compliant?  If not, if this is a well-known exception, it might be useful to others.  OWSLib an OGC sponsored project, so it may be appropriate to add a "strict=True" flag that is set by default.  If strict=False, it will make these non-conformant WaterML files work with OWSLib?  Otherwise, if an OGC standard (WaterML) cannot be read by OGC-sponsored software (OWSLib), why have our colleagues (some of whom are on the HydroShare project) bothered with the standard in the first place?

zhiyuli commented:
@selimnairb I removed all modified OWSLib source code from our repo. Now ref_ts only references an official OWSLib release. Should we include its license in our repo?  @mjstealey Is there doc in our repo describing the licenses of all third party libraries we are referencing? 

also I changed to use an existing logger.

Will you please continue reviewing it? thanks

selimnairb commented on ref_ts/tests/test_ref_ts.py:
```diff
@@ -0,0 +1,239 @@
+        self.url_to_search_variables = "/hsapi/_internal/search-variables/"
+        self.url_to_time_series_from_service = "/hsapi/_internal/time-series-from-service/"
+        self.url_to_create_ref_time_series = "/hsapi/_internal/create-ref-time-series/"
+        self.url_to_download_resource_files = "/hsapi/_internal/{0}/download-files/"
```
Can you use /hsapi/resource/<pk>/file/<filename>/ for this?  If so, then we should get rid of this view as it duplicates what is in the REST API.

selimnairb commented on ref_ts/views.py:
```diff
@@ -1,222 +1,285 @@
 
-class ReferencedSitesForm(forms.Form):
-    url = forms.URLField()
+preview_name = "preview.png"
```
Capitalize variable names for constants.

selimnairb commented on ref_ts/views.py:
```diff
@@ -1,222 +1,285 @@
-class ReferencedSitesForm(forms.Form):
-    url = forms.URLField()
+preview_name = "preview.png"
+his_central_url = 'http://hiscentral.cuahsi.org/webservices/hiscentral.asmx/GetWaterOneFlowServiceInfo'
```
Capitalize variable names for constants.

selimnairb commented on ref_ts/views.py:
```diff
@@ -1,222 +1,285 @@
+preview_name = "preview.png"
+his_central_url = 'http://hiscentral.cuahsi.org/webservices/hiscentral.asmx/GetWaterOneFlowServiceInfo'
+logger = logging.getLogger("django")
+blank_field_string = ""
```
Capitalize variable names for constants.

selimnairb commented:
@zhiyuli Sorry for the delay.  I have finished reviewing this PR and it looks good.  Before merging we need to address the following:
  - Replace and remove download_resource_files view with use of REST API (unless download_resource_files view is doing something very different)
  - Capitalize variable names for constants

zhiyuli commented:
@selimnairb 
Thanks for your time reviewing this. 
The ref_ts is different from other resource types such that it does not store resource files on irods (there is a bag on irods for every ref_ts resource but it only contains xml files recording metadata). 
We replace the original "download button" in generic.html with a "ref_ts download button". The new button points to url "/hsapi/_internal/{res_id}/download-files/", which links to function ref_ts.views.download_resource_files().

This function is doing the following things:
1) check the caller's (user) permission on the requested ref_ts resource.
2) download the bag from irods
3) download time series data (wml) by querying the referenced url stored in metadata obj
4) merge time series data into the bag (add resource files)
5) return the filled bag to front end.

is there a way that we can include this implementation into existing rest api? 
Thanks

selimnairb commented on ref_ts/views.py:
```diff
@@ -1,222 +1,285 @@
+    except Exception as e:
+        logger.exception("download_resource_files: %s" % (e.message))
+        response = HttpResponse(status=503)
+        response.content = "<h3>Failed to download this resource!</h3>"
```
There should not be any HTML or other markup in our views.

selimnairb commented:
@zhiyuli I understand now (re. file download).  I think this is specific enough to this resource type that it needs to be a separate API call.  However, if we think others might want to use this view, then the URL  should not be *_internal/\<pk\>/download-files/* but instead something like *hsapi/RefTimeSeriesResource/\<pk\>/downloadFiles/*.

@horsburgh How do you think we should organize REST API end points provided by specific resource types?

@zhiyuli Another thing to think about for a future minor revision is what happens if the bag download process take more than a couple of seconds?  In this case, we might need to do this asynchronously using a celery task.

horsburgh commented:
@selimnairb and @zhiyuli - I was hoping that there wouldn't be different API methods for each resource type. Could we handle this via code in the API that checks the resource type for a requested resource and does the appropriate action to return the resource files rather than having a separate API method for every resource type (or at least each resource type that requires something different)?  In other words, keep the API interface consistent, but handle the cases on a type by type basis on the server before returning a response?

Another approach (that is perhaps less satisfying, but wouldn't require changes to the API) is that HydroShare returns the resource exactly as it is in HydroShare - without the WaterML file(s) and it is up to the user to follow the links in the resource metadata.

zhiyuli commented:
@selimnairb 
I addressed the comment about naming constant variables. Also I renamed the download url to hsapi/RefTimeSeriesResource/<pk>/downloadFiles/ as it is only meaningful to ref_ts. 

@horsburgh  @selimnairb  about the rest api,  i think we can keep as is (at least for short term):  
1) on the ref_ts landing page, user can download a "updated" bag that contains wml, csv and preview figure.
2) if using rest api,  just return the original bag in irods. users can get wml by themselves.

danames commented:
@zhiyuli I think this is a good solution. @horsburgh is right - we don't
want to complicate the API with special case methods for different resource
types.

On Mon, Feb 1, 2016 at 5:26 PM Drew Li <notifications@github.com> wrote:

> @selimnairb <https://github.com/selimnairb>
> I addressed the comment about naming constant variables. Also I renamed
> the download url to hsapi/RefTimeSeriesResource//downloadFiles/ as it is
> only meaningful to ref_ts.
>
> @horsburgh <https://github.com/horsburgh> @selimnairb
> <https://github.com/selimnairb> about the rest api, i think we can keep
> as is (at least for short term):
>
> 1) on the ref_ts landing page, user can download a "updated" bag that
> contains wml, csv and preview figure.
> 2) if using rest api, just return the original bag in irods. users can get
> wml by themselves.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hydroshare/hydroshare/pull/882#issuecomment-178271673>
> .
>

zhiyuli commented:
@selimnairb do you think this PR is ready to go? thanks

selimnairb commented:
@zhiyuli Unfortunately, I think renaming the API end point to "hsapi/RefTimeSeriesResource//downloadFiles/" [sic] contradicts @horsburgh and @danames wish to no add any resource-specific end points.  I don't want to hold up this PR any longer.  For now, I would revert the name of the download end point to what it was previously (i.e. _internal ...; a naming convention that I hate, but we are stuck with for the present).  After this is done, and the branch is brought up to date with develop, then it can be merged.  We can look at bringing the RefTs download feature into the REST API proper in a subsequent issue.

zhiyuli commented:
@selimnairb I changed the api end point back to "_internal" and sync'd with dev. thanks

@mjstealey this pr is ready, please merge it once it passes jenkins. thanks
